subprojects/signing/src/main/java/org/gradle/plugins/signing/SigningExtension.java

```diff
@@ -408,15 +408,21 @@ private Sign createSignTaskFor(CharSequence name, Action<Sign> taskConfiguration
         return signTask;
     }
 
-    protected Object addSignaturesToConfiguration(Sign task, final Configuration configuration) {
-        task.getSignatures().all(new Action<Signature>() {
-            public void execute(Signature sig) {
-                configuration.getArtifacts().add(sig);
-            }
-        });
-        return task.getSignatures().whenObjectRemoved(new Action<Signature>() {
-            public void execute(Signature sig) {
-                configuration.getArtifacts().remove(sig);
+    protected void addSignaturesToConfiguration(Sign task, final Configuration configuration) {
+        project.afterEvaluate(new Action<Project>() {
```

Comment:
We want to get rid of afterEvaluate, not introduce it more. One of the goals when we did stable publishing was relying on afterEvaluate as little as possible.

If the lazy task API forces us into sprinkling afterEvaluate, then I think something is off there.

Reply:
This really doesn't have anything to do with the lazy tasks.  It's due to the circular relationship of rules between the signature plugin, the archives configuration, and `DefaultArtifactPublicationSet`.  We've introduced some mutation validation on domain object collections in 5.0 that is now catching the resulting double-add of the signature.  I was trying to avoid this blowing up into a redesign of the signature plugin, hence the hack.  I'll see if there is something else simple we can do here.

Reply:
I see, I thought we had added that validation just for tasks. 

> catching the resulting double-add of the signature

Sounds like we should have some code that notices that the added artifact is a signature and break the cycle at that point.

Reply:
Yeah, that's pretty much what I've done - we did already check for the signature artifact, but we didn't check for adding a signature for the same artifact twice (which is ultimately what was happening because of the circular rules).  Previously, this was ok (and worked for the wrong reasons), but with the mutation checks, we were now catching the fact that we were doing an add inside of an add.  We're exiting this mad cycle earlier now and avoiding the extra signature.